### PR TITLE
feat: ignore Flutter builder functions in "prefer-extracting-callbacks" rule

### PR DESCRIPTION
**Pull request by roman-petrov**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[x] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

### What changes did you make? (Give an overview)

It is a pretty common situation in Flutter to have "builders". Widgets even may have multiple "builder" arguments like `titleBuilder`, `contentBuilder`, etc. We are getting violations from dart code metrics on such arguments so I think it would be ok to ignore such arguments in `prefer-extracting-callbacks` rule.

### Is there anything you'd like reviewers to focus on?

I will create a separate PR to website repository with documentation updates.  It's a little bit annoying to manually keep this repository documentation and website documentation synchronized, maybe we should start thinking about avoiding this? (we can define primary documentation source somehow, for example)

### CODE REVIEW

**Comment by dkrutskikh** [comment]
@roman-petrov I think detect `builder` only by name it's not a good idea. What if we check lambda function arguments and find argument with `BuildContex` type?

**Comment by roman-petrov** [comment]
> @roman-petrov I think detect `builder` only by name it's not a good idea. What if we check lambda function arguments and find argument with `BuildContex` type?

Thank you, I think it might be a good idea to try to identify builders by function argument type. So we can, for example, assume that the argument is a "builder" if its RHS expression is a function with the first argument of type `BuildContext`. So, again, 
the heuristic can be described as **Ignore all arguments which value is a function with the first argument having type `BuildContext`**

In this case we can even avoid checking argument names (we currently ignore arguments named "builder"). So will it be ok if we totally remove the argument name check and ignore "builders" based on the heuristics above, what do you think? 

 @incendial , @dkrutskikh

**Comment by incendial** [comment]
@roman-petrov sorry for the long response from my side, I'll take a closer look on weekends.

**Comment by incendial** [comment]
@roman-petrov I'm not a fan of ignoring several arguments by their name, but the idea about checking for BuildContext looks solid.

**Comment by roman-petrov** [comment]
@incendial , thank you, I will proceed with the new approach.

**Comment by roman-petrov** [comment]
I am done with this, PTAL.

File: lib/src/analyzers/lint_analyzer/rules/rules_list/prefer_extracting_callbacks/visitor.dart [hunk 1]
```diff
@@ -55,8 +56,19 @@ class _InstanceCreationVisitor extends RecursiveAstVisitor<void> {
     return body.block.statements.isNotEmpty;
   }
 
+  bool _isFlutterBuilder(FunctionExpression expression) {
+    if (!isWidget(expression.declaredElement?.returnType)) {
```
Comment (dkrutskikh) [inline lib/src/analyzers/lint_analyzer/rules/rules_list/prefer_extracting_callbacks/visitor.dart:60]: @roman-petrov in some cases builder function returns subclass of widget, for example:

https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/cupertino/text_field.dart#L1054 , in this case builder returns Row classю

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

File: lib/src/analyzers/lint_analyzer/rules/rules_list/prefer_extracting_callbacks/visitor.dart (hunk 1 above)
Comment (roman-petrov) [reply lib/src/analyzers/lint_analyzer/rules/rules_list/prefer_extracting_callbacks/visitor.dart]: Done, please take a look to updated code/test/documentation.

**Review by dkrutskikh — APPROVED** [review]

**Comment by dkrutskikh** [comment]
@roman-petrov now all looks good for me, thank you

**Pull request merged**